src/main/kotlin/com/example/delivery/service/ProductReserveService.kt

```diff
@@ -0,0 +1,30 @@
+package com.example.delivery.service
+
+import com.example.delivery.domain.DomainProduct
+import org.springframework.stereotype.Service
+import java.math.BigDecimal
+
+@Service
+class ProductReserveService {
+    fun reserveProducts(items: Map<DomainProduct, Int>) {
+        items.forEach { (product, amount) ->
+            validateProductAvailability(product, amount)
+        }
+    }
+
+    private fun validateProductAvailability(product: DomainProduct, amount: Int) {
+        val availableAmount = product.amountAvailable
+        if (availableAmount < amount) {
+            throw ArithmeticException("Not enough items for product ${product.id}")
```

Comment:
Create a custom exception. 

From `ArithmeticException` javadocs:
```
 * Thrown when an exceptional arithmetic condition has occurred. For
 * example, an integer "divide by zero" throws an
 * instance of this class.
```

The exception is about arithmetic errors, not about business logic